src/providers/ldap/ldap_auth.c

```diff
@@ -1369,29 +1390,53 @@ static void sdap_pam_chpass_handler_auth_done(struct tevent_req *subreq)
-                DEBUG(SSSDBG_CRIT_FAILURE,
-                      "Changing shadow password attributes not implemented.\n");
-                state->pd->pam_status = PAM_MODULE_UNKNOWN;
-                goto done;
-            } else {
-                subreq = sdap_pam_change_password_send(state, state->ev,
-                                                       state->sh,
-                                                       state->auth_ctx->opts,
-                                                       state->pd,
-                                                       state->dn);
-                if (subreq == NULL) {
-                    DEBUG(SSSDBG_OP_FAILURE, "Failed to change password for "
-                          "%s\n", state->pd->user);
-                    state->pd->pam_status = PAM_SYSTEM_ERR;
-                    goto done;
-                }
 
-                tevent_req_set_callback(subreq,
-                                        sdap_pam_chpass_handler_chpass_done,
-                                        req);
-                return;
+                update_option_set_explictly = confdb_is_set_explicit(
+                                                       state->be_ctx->cdb,
+                                                       state->be_ctx->conf_path,
+                                                       state->auth_ctx->opts->basic[SDAP_CHPASS_UPDATE_LAST_CHANGE].opt_name);
+
+                if (!dp_opt_get_bool(state->auth_ctx->opts->basic,
+                                     SDAP_CHPASS_UPDATE_LAST_CHANGE)
+                            && !update_option_set_explictly) {
+                    DEBUG(SSSDBG_CRIT_FAILURE,
+                       "Shadow password policy is selected but "
+                       "ldap_chpass_update_last_change is not set, please "
+                       "make sure your LDAP server can update the [%s] "
+                       "attribute automatically. Otherwise SSSD might "
+                       "consider your password as expired.\n",
+                       state->auth_ctx->opts->user_map[SDAP_AT_SP_LSTCHG].name);
+                }
+                ret = sysdb_invalidate_cache_entry(
+                                               state->be_ctx->domain,
+                                               state->pd->user, true);
+                if (ret != EOK) {
+                    DEBUG(SSSDBG_OP_FAILURE,
+                       "Failed to invalidate cache entry for user [%s]. The "
+                       "last changed attribute [%s] might not get refreshed "
+                       "after the password and the password might still be "
+                       "considered as expired. Call sss_cache for this user "
+                       "to expire the entry manually in this case.\n",
+                       state->pd->user,
+                       state->auth_ctx->opts->user_map[SDAP_AT_SP_LSTCHG].name);
+                }
```

Comment:
Implementing the complete `ERR_PASSWORD_EXPIRED` case in a function or at least the selected part would make it easier to read and maintain. Moreover, it would help you stick to 80 characters per line.

Edit: apparently this was already pointed out by CodeQL: https://github.com/SSSD/sssd/security/code-scanning/1190